src/main/java/net/snowflake/ingest/streaming/internal/ParquetFlusher.java

```diff
@@ -224,4 +230,64 @@ private SerializationResult serializeFromJavaObjects(
         mergedData,
         chunkMinMaxInsertTimeInMs);
   }
+
+  /**
+   * Validates that rows count in metadata matches the row count in Parquet footer and the row count
+   * written by the parquet writer
+   *
+   * @param serializationType Serialization type, used for logging purposes only
+   * @param writer Parquet writer writing the data
+   * @param channelsDataPerTable Channel data
+   * @param javaSerializationTotalRowCount Total row count when java object serialization is used.
```

Comment:
What do you mean by `java object serialization is used` - looks like above this is just the total row count?

Reply:
The SDK implements two types of row buffering. The default one, which buffers rows as lists of java objects in memory and during flush it iterates over all of them and writes them to Parquet. There is also an alternative implementation called internal buffering implemented by @sfc-gh-azagrebin, which doesn't collect any Java values, but which instead directly serializes rows to Parquet. So, for the default buffering method (java objects), we have 4 sources of row counts that all have to match:

1. size of the buffered `List<List<Object>>`
2. number of rows written by the Parquet writer
3. number of rows in Parquet footer
4. number of rows collected for metadata

For the internal buffering optimization, we only have 2nd, 3rd and 4th.